plugins/module_utils/common.py

```diff
@@ -98,112 +99,83 @@
     K8S_IMP_ERR = traceback.format_exc()
 
 
-def list_dict_str(value):
-    if isinstance(value, (list, dict, string_types)):
-        return value
-    raise TypeError
-
-
-ARG_ATTRIBUTES_BLACKLIST = ('property_path',)
-
-COMMON_ARG_SPEC = {
-    'state': {
-        'default': 'present',
-        'choices': ['present', 'absent'],
-    },
-    'force': {
-        'type': 'bool',
-        'default': False,
-    },
-}
-
-RESOURCE_ARG_SPEC = {
-    'resource_definition': {
-        'type': list_dict_str,
-        'aliases': ['definition', 'inline']
-    },
-    'src': {
-        'type': 'path',
-    },
-}
-
-NAME_ARG_SPEC = {
-    'kind': {},
-    'name': {},
-    'namespace': {},
-    'api_version': {
-        'default': 'v1',
-        'aliases': ['api', 'version'],
-    },
-}
-
-AUTH_ARG_SPEC = {
-    'kubeconfig': {
-        'type': 'path',
-    },
-    'context': {},
-    'host': {},
-    'api_key': {
-        'no_log': True,
-    },
-    'username': {},
-    'password': {
-        'no_log': True,
-    },
-    'validate_certs': {
-        'type': 'bool',
-        'aliases': ['verify_ssl'],
-    },
-    'ca_cert': {
-        'type': 'path',
-        'aliases': ['ssl_ca_cert'],
-    },
-    'client_cert': {
-        'type': 'path',
-        'aliases': ['cert_file'],
-    },
-    'client_key': {
-        'type': 'path',
-        'aliases': ['key_file'],
-    },
-    'proxy': {
-        'type': 'str',
-    },
-    'persist_config': {
-        'type': 'bool',
-    },
-}
-
-WAIT_ARG_SPEC = dict(
-    wait=dict(type='bool', default=False),
-    wait_sleep=dict(type='int', default=5),
-    wait_timeout=dict(type='int', default=120),
-    wait_condition=dict(
-        type='dict',
-        default=None,
-        options=dict(
-            type=dict(),
-            status=dict(type='str', default="True", choices=["True", "False", "Unknown"]),
-            reason=dict()
-        )
-    )
-)
-
-# Map kubernetes-client parameters to ansible parameters
-AUTH_ARG_MAP = {
-    'kubeconfig': 'kubeconfig',
-    'context': 'context',
-    'host': 'host',
-    'api_key': 'api_key',
-    'username': 'username',
-    'password': 'password',
-    'verify_ssl': 'validate_certs',
-    'ssl_ca_cert': 'ca_cert',
-    'cert_file': 'client_cert',
-    'key_file': 'client_key',
-    'proxy': 'proxy',
-    'persist_config': 'persist_config',
-}
+def get_api_client(module=None, **auth_params):
+    auth_params = auth_params or {}
+    auth = {}
+
+    import hashlib
+    m = hashlib.sha256()
+    for i in AUTH_ARG_MAP:
+        m.update(str(auth_params.get(i)).encode())
+    digest = m.hexdigest()
+    if digest in get_api_client._pool:
+        client = get_api_client._pool[digest]
+        return client
```

Comment:
I don't think we can correctly return here, as it's possible that the content of the kubeconfig, the content of the incluster config, or the envvars could have changed since the last time this was called. I think we need to wait until we've pulled the variables from all possible locations to create this digest. IOW, the kubernetes.client.Configuration object would ideally be the object that is used to create this digest.

Reply:
thank you @fabianvf for the review. I agree, it's a much better strategy. I just pushed a new revision of the patch that uses this strategy.